### ISSUE TITLE
The lifecycle of ObservableQuery with recycling is broken

### ISSUE THREAD

**Issue by jamesreggio** [apollographql/apollo-client#2196]
First, thank you for the excellent work on 2.0.

We're seeing upwards of a 5x perf improvement in our React Native app! 🎉

**What's wrong**

One strange regression I've witnessed pertains to the interactions between the query recycler in `react-apollo` and the new `ObservableQuery`. The `react-apollo` recycler will retain and reuse the same `ObservableQuery` for each instance of the component it wraps. (So far, so good. Yay perf.) However, this `ObservableQuery` will return data for its prior result, even after the variables supplied to the query have changed.

More concretely, say you have an `EpisodeDetailsScreen` with an `EpisodeDetailsQuery`. The query accepts an `episodeId` variable. Every time you navigate to the `EpisodeDetailsScreen`, the `react-apollo` recycler will reuse the same instance of the `ObservableQuery`.

Say you navigate to `EpisodeDetailsScreen` for `episodeId: 10`, and then you navigate away. Then you navigate to `EpisodeDetailsScreen` for `episodeId: 20`. The `ObservableQuery` will be recycled for this second navigation, but it will return the results for `episodeId: 10` initially, until the results for `episodeId: 20` have been loaded from the network.

(This is for `fetchPolicy: 'cache-and-network'`, though I haven't confirmed that it doesn't repro on other `fetchPolicies`.)

**Suggested fix**

The simplest solution is to clear the values of `lastResult` and `lastError` on `ObservableQuery` after new variables have been supplied to `setVariables`. This is the local workaround I'm using.

However, the reason I didn't submit a PR is that there's a field on `ObservableQuery` called `isTornDown` whose purpose is unclear to me.

If you look carefully, you'll see that once `isTornDown` is set to `true`, it can never been reverted to `false`. This seems to stand at odds with the idea that an `ObservableQuery` can be recycled. Furthermore, once `true`, the logic in `currentResult` is always short-circuited. (`react-apollo` uses `currentResult` to obtain the latest data in each render pass.)

Because I'm not aware of the design considerations that went into the `isTornDown` bookkeeping, I figured I'd let y'all figure out the best course of action. May I suggest that whatever the solution, an assertion is put into place to prevent calls to methods (like `setVariables`) that should be illegal for torn-down queries.

**Version**

- apollo-client@beta
- react-apollo@1.4.15

**Comment by affablebloke** [comment]
We have been experiencing the exact same issue.

**Comment by jbaxleyiii** [comment]
@affablebloke is that with the beta or the current 1.* branch?

**Comment by affablebloke** [comment]
@jbaxleyiii 
```
    "apollo-cache-inmemory": "beta",
    "apollo-client": "beta",
    "apollo-link": "^0.5.4",
    "apollo-link-http": "beta",
    "react-apollo": "^1.4.15",
```

BTW I love apollo-link, it is such a great design decision.

**Comment by davemcorwin** [comment]
I'm having this same issue, does anyone have a suggested workaround or a way to force the variables NOT to be reused?

I noticed there is a `storePreviousVariables` key in the Redux action, is there anyway to control that?

react-apollo 1.4.15
apollo-client 1.9.3
apollo-link-core 0.5.4

**Comment by jamesreggio** [comment]
This bug should actually be fixed in the latest version of react-apollo and
apollo-client pushed yesterday. Just reinstall everything at the @beta tag
and see if that fixes it for you.

**Comment by davemcorwin** [comment]
Thanks @jamesreggio is there a solution that doesn't involve bumping a major version, there are breaking changes here.

**Comment by jbaxleyiii** [comment]
@davemcorwin the breaking changes only impact one file! Is there something large preventing you from upgrading?

**Comment by davemcorwin** [comment]
@jbaxleyiii I will look at https://github.com/apollographql/apollo-client/blob/master/Upgrade.md, but I have a production application, so doing an off-the-cuff major version upgrade to a beta makes me a bit nervous...

**Comment by jbaxleyiii** [comment]
@davemcorwin of course! I totally understand! If you can hold off for one more week it will be the official release! Its testing super well and we just have doc updates left with full upgrade guides before cutting the release candidate and moving to an official release

**Comment by AndreasGalster** [comment]
Could this issue be related to updateQueries not properly updating the observableQuery as well?

**Comment by jbaxleyiii** [comment]
This is fixed in the 2.0. If someone wants to open a PR for the latest I'd be happy to merge and release it but the official recommendation is to upgrade to the 2.0
